omniscidb/QueryEngine/Compiler/HelperFunctions.cpp

```diff
@@ -101,15 +104,88 @@ void eliminate_dead_self_recursive_funcs(
 }
 #endif
 
+std::string legalizePassName(const llvm::StringRef pass_name) {
+  std::string legal_name = pass_name.str();
+  constexpr std::string_view illegal_word = "w/";
+  const size_t illegal_word_pos = legal_name.find(illegal_word);
+  if (illegal_word_pos != std::string::npos) {
+    std::string replacement =
+        "with" + legal_name.substr(illegal_word_pos + illegal_word.size());
+    legal_name.replace(illegal_word_pos, replacement.size(), replacement);
+  }
+  return legal_name;
+}
+
+const llvm::Module* unwrapModule(llvm::Any IR) {
+  if (const auto** M = llvm::any_cast<const llvm::Module*>(&IR))
+    return *M;
+
+  if (const auto** F = llvm::any_cast<const llvm::Function*>(&IR)) {
+    return (*F)->getParent();
+  }
+
+  if (const auto** C = llvm::any_cast<const llvm::LazyCallGraph::SCC*>(&IR)) {
+    for (const llvm::LazyCallGraph::Node& N : **C) {
+      const llvm::Function& F = N.getFunction();
+      return F.getParent();
+    }
+  }
+
+  if (const auto** L = llvm::any_cast<const llvm::Loop*>(&IR)) {
+    const llvm::Function* F = (*L)->getHeader()->getParent();
+    return F->getParent();
+  }
+  return nullptr;
+}
+
+std::string getStrTStamp() {
+  char t_stamp_str[100];
+  const std::time_t time = std::chrono::system_clock::to_time_t(work_unit_meta.t_stamp);
+  const auto t_stamp_ms = std::chrono::duration_cast<std::chrono::milliseconds>(
+      work_unit_meta.t_stamp.time_since_epoch());
+  std::strftime(t_stamp_str, sizeof(t_stamp_str), "%F_%T", std::localtime(&time));
+  return std::string{t_stamp_str}.append(".").append(
+      std::to_string(t_stamp_ms.count() % 1000));
+}
+
 void optimize_ir(llvm::Function* query_func,
                  llvm::Module* llvm_module,
                  const std::unordered_set<llvm::Function*>& live_funcs,
                  const bool is_gpu_smem_used,
                  const CompilationOptions& co) {
   auto timer = DEBUG_TIMER(__func__);
 
-  llvm::PassBuilder PB;
+  llvm::PassInstrumentationCallbacks PIC;
+  size_t pass_counter{1};
+  std::string ir_dump_dir{};
+  if (co.dump_llvm_ir_after_each_pass) {
+    work_unit_meta.w_unit_counter++;
```

Comment:
What's going to happen in the heterogeneous mode when you potentially have two compilations running simultaneously? Should the `work_unit_meta` live in the work unit?

Reply:
Yes, the timestamp can be saved in the module for debug purposes and then you can ask for it in `optimize_ir()`. But there are lots of ways to get into `optimize_ir()`, so where do you embed the timestamp into the module?
E.g., you can embed the timestamp in `Executor::compileWorkUnit()`, but CPU codegen path might go through `Executor::reduceMultiDeviceResultSets() -> ResultSetReductionJIT::codegen() -> ...`, omitting `Executor::compileWorkUnit()`, so how many other places should have code that embeds the timestamp? Right now it seems to me that this is the least amount of code that does the job, or am I missing something?